crates/re_smart_channel/src/lib.rs

```diff
@@ -75,29 +159,71 @@ fn smart_channel_with_stats<T: Send>(
     (sender, receiver)
 }
 
-/// Stats for a channel, possibly shared between chained channels.
-#[derive(Default)]
-struct SharedStats {
-    /// Latest known latency from sending a message to receiving it, it nanoseconds.
-    latency_ns: AtomicU64,
-}
+// ---
 
 #[derive(Clone)]
 pub struct Sender<T: Send> {
-    tx: crossbeam::channel::Sender<(Instant, T)>,
+    tx: crossbeam::channel::Sender<SmartMessage<T>>,
+    source: Arc<SmartMessageSource>,
     stats: Arc<SharedStats>,
 }
 
 impl<T: Send> Sender<T> {
+    /// Clones the sender with an updated source.
+    pub fn clone_as(&self, source: SmartMessageSource) -> Self {
+        Self {
+            tx: self.tx.clone(),
+            source: Arc::new(source),
+            stats: Arc::clone(&self.stats),
+        }
+    }
+
     pub fn send(&self, msg: T) -> Result<(), SendError<T>> {
-        self.send_at(Instant::now(), msg)
+        self.send_at(
+            Instant::now(),
+            Arc::clone(&self.source),
+            SmartMessagePayload::Msg(msg),
+        )
     }
 
-    /// back-date a message
-    pub fn send_at(&self, time: Instant, msg: T) -> Result<(), SendError<T>> {
+    /// Forwards a message as-is.
+    pub fn send_at(
+        &self,
+        time: Instant,
+        source: Arc<SmartMessageSource>,
+        payload: SmartMessagePayload<T>,
+    ) -> Result<(), SendError<T>> {
+        // NOTE: We should never be sending a message with an unknown source.
+        debug_assert!(!matches!(*source, SmartMessageSource::Unknown));
+
         self.tx
-            .send((time, msg))
-            .map_err(|SendError((_, msg))| SendError(msg))
+            .send(SmartMessage {
+                time,
+                source,
+                payload,
+            })
+            .map_err(|SendError(msg)| SendError(msg.into_data().unwrap()))
+    }
+
+    /// Used to indicate that a sender has left.
+    ///
+    /// This sends a message down the channel allowing the receiving end to know whether one of the
+    /// sender has left, and if so why (if applicable).
+    ///
+    /// Using a [`Sender`] after calling `quit` is undefined behaviour: the receiving end is free
+    /// to silently drop those messages (or worse).
+    pub fn quit(
+        &self,
```

Comment:
Can this take `self`?

Reply:
Not really, we need to use this in all sorts of weird contexts, e.g. `Fn` (rather than `FnOnce`) callbacks :(